### PR TITLE
vulkaninfo: Add checks for some struct allocation

### PR DESCRIPTION
**Pull request by jeremyk-lunarg**
Added checks after attempted memory allocation in
AppGpuInit and before memory frees in
AppGpuDestroy to fix crashing.

### CODE REVIEW

File: demos/vulkaninfo.c [hunk 1]
```diff
@@ -956,14 +964,26 @@ static void AppGpuInit(struct AppGpu *gpu, struct AppInstance *inst, uint32_t id
...
-    free(b16_store_features);
+    if (CheckExtensionEnabled(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME, gpu->inst->inst_extensions,
+        gpu->inst->inst_extensions_count)) {
+        VkPhysicalDevice16BitStorageFeaturesKHR *b16_store_features = gpu->features2.pNext;
+        VkPhysicalDeviceSamplerYcbcrConversionFeaturesKHR *ycbcr_conv_features = b16_store_features->pNext;
+        VkPhysicalDeviceVariablePointerFeaturesKHR *var_pointer_features = ycbcr_conv_features->pNext;
+        VkPhysicalDeviceBlendOperationAdvancedFeaturesEXT *blend_op_adv_features = var_pointer_features->pNext;
+
+        if (blend_op_adv_features) {
+            free(blend_op_adv_features);
```
Comment (polarina) [inline demos/vulkaninfo.c]: The `if`s here are not needed. The `free` function is a guaranteed no-op on NULL pointers.

File: demos/vulkaninfo.c [hunk 2]
```diff
@@ -832,18 +832,22 @@ static void AppGpuInit(struct AppGpu *gpu, struct AppInstance *inst, uint32_t id
                               gpu->inst->inst_extensions_count)) {
         gpu->props2.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2_KHR;
         gpu->props2.pNext = malloc(sizeof(VkPhysicalDeviceBlendOperationAdvancedPropertiesEXT));
+        assert(gpu->props2.pNext);
```
Comment (jzulauf-lunarg) [inline demos/vulkaninfo.c]: ERR_EXIT(VK_ERROR_OUT_OF_HOST_MEMORY);

would be more consistent with the rest of the setup allocations (see main -- where this is called from)

File: demos/vulkaninfo.c [hunk 3]
```diff
@@ -956,14 +964,18 @@ static void AppGpuInit(struct AppGpu *gpu, struct AppInstance *inst, uint32_t id
...
-    VkPhysicalDeviceSamplerYcbcrConversionFeaturesKHR *ycbcr_conv_features = b16_store_features->pNext;
-    VkPhysicalDeviceVariablePointerFeaturesKHR *var_pointer_features = ycbcr_conv_features->pNext;
-    VkPhysicalDeviceBlendOperationAdvancedFeaturesEXT *blend_op_adv_features = var_pointer_features->pNext;
-    free(blend_op_adv_features);
-    free(var_pointer_features);
-    free(ycbcr_conv_features);
-    free(b16_store_features);
+    if (CheckExtensionEnabled(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME, gpu->inst->inst_extensions,
+        gpu->inst->inst_extensions_count)) {
+        VkPhysicalDevice16BitStorageFeaturesKHR *b16_store_features = gpu->features2.pNext;
```
Comment (jzulauf-lunarg) [inline demos/vulkaninfo.c]: wouldn't it be more general to run while loop over the pNext chain freeing the node after stashing away a copy of pNext for the next iteration?  You could use it here and down at line 990

**Comment by mikew-lunarg** [comment]
FWIW attempting to test this PR w Devsim from VulkanTools, but getting bogged down in updating the LVL submodules.  Will report ASAP.

File: demos/vulkaninfo.c (hunk 2 above)
Comment (jeremyk-lunarg) [reply demos/vulkaninfo.c]: Fixed.

File: demos/vulkaninfo.c (hunk 3 above)
Comment (jeremyk-lunarg) [reply demos/vulkaninfo.c]: Added this.  Good idea to help keep the code maintainable.

Comment (jzulauf-lunarg) [reply demos/vulkaninfo.c]: I was thinking looping in the "free" sections....

Comment (jeremyk-lunarg) [reply demos/vulkaninfo.c]: Whoops.  Fixed it.

File: demos/vulkaninfo.c [hunk 4]
```diff
@@ -956,14 +981,17 @@ static void AppGpuInit(struct AppGpu *gpu, struct AppInstance *inst, uint32_t id
...
-    VkPhysicalDeviceVariablePointerFeaturesKHR *var_pointer_features = ycbcr_conv_features->pNext;
-    VkPhysicalDeviceBlendOperationAdvancedFeaturesEXT *blend_op_adv_features = var_pointer_features->pNext;
-    free(blend_op_adv_features);
-    free(var_pointer_features);
-    free(ycbcr_conv_features);
-    free(b16_store_features);
+    if (CheckExtensionEnabled(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME, gpu->inst->inst_extensions,
+                              gpu->inst->inst_extensions_count)) {
+        struct VkStructureHeader *place = gpu->features2.pNext;
+        struct VkStructureHeader *next = NULL;
```
Comment (jzulauf-lunarg) [inline demos/vulkaninfo.c]: this is kind of duplicate code to put in a small utility function.

File: demos/vulkaninfo.c [hunk 5]
```diff
@@ -974,16 +1002,16 @@ static void AppGpuDestroy(struct AppGpu *gpu) {
...
 
-    VkPhysicalDeviceBlendOperationAdvancedPropertiesEXT *blend_op_adv_props = gpu->props2.pNext;
-    VkPhysicalDevicePointClippingPropertiesKHR *pt_clip_props = blend_op_adv_props->pNext;
-    VkPhysicalDevicePushDescriptorPropertiesKHR *push_desc_props = pt_clip_props->pNext;
-    VkPhysicalDeviceDiscardRectanglePropertiesEXT *discard_rect_props = push_desc_props->pNext;
-    free(discard_rect_props);
-    free(push_desc_props);
-    free(pt_clip_props);
-    free(blend_op_adv_props);
+        struct VkStructureHeader *place = gpu->props2.pNext;
```
Comment (jzulauf-lunarg) [inline demos/vulkaninfo.c]: Not very DRY :-1:  to repeat

Comment (jeremyk-lunarg) [reply demos/vulkaninfo.c]: Fixed it to make it less WET.

**Comment by mikew-lunarg** [comment]
Verified that this fixes the crash that inspired this thread 👍

**Comment by jeremyk-lunarg** [comment]
Since I have addressed every suggestion so far, the PR is passing all CI testing, and @mikew-lunarg needs this for his project, if there are no further comments on this by 11AM tomorrow (April 4) I will merge the PR.

File: demos/vulkaninfo.c [hunk 6]
```diff
@@ -484,6 +484,36 @@ static bool CheckPhysicalDeviceExtensionIncluded(const char *extension_to_check,
...
+            }
+        } else {
+            place->pNext = NULL;
+        }
+
+        place = place->pNext;
+    }
+}
+
+static void freepNextChain(struct VkStructureHeader *first) {
```
Comment (jzulauf-lunarg) [inline demos/vulkaninfo.c]: :1st_place_medal:

**Pull request merged**